### PR TITLE
Fixes #30466 - trigger toasts from API middleware

### PR DESCRIPTION
**Pull request by Ron-Lavi**
## Toast handling

Instead of managing and dispatching the toast action by yourself, you can use the API middleware shortcuts,
by passing `toastSuccess` or `toastError` as keys in the API action, e.g:

```js
dispatch(
  get({
    key: MY_SPECIAL_KEY,
    url,
    toastError: error => `Oh no! Something went wrong while submiting the form, the server returned the following error: ${error}`,
    toastSuccess: response => `Yay! task: ${response.taskID} was successfully created.`
  })
);
```

c7c586875109ec38f103e3fc9985c0b2

### CODE REVIEW

File: webpack/assets/javascripts/react_app/redux/API/APIRequest.js [hunk 1]
```diff
@@ -15,26 +16,37 @@ export const apiRequest = async (
...
+      successToast,
+      errorToast,
       payload = {},
     },
   },
   { dispatch, getState }
 ) => {
   const { REQUEST, SUCCESS, FAILURE } = actionTypeGenerator(key, actionTypes);
   const modifiedPayload = { ...payload, url };
+
```
Comment (Ron-Lavi) [inline webpack/assets/javascripts/react_app/redux/API/APIRequest.js:28]: Added some spaces for readability

File: webpack/assets/javascripts/react_app/redux/actions/toasts/index.js [hunk 2]
```diff
@@ -3,7 +3,7 @@ import uuidV1 from 'uuid/v1';
 import { TOASTS_ADD, TOASTS_DELETE, TOASTS_CLEAR } from '../../consts';
 
 export const addToast = toast => {
-  const key = uuidV1();
+  const key = toast.key || uuidV1();
```
Comment (Ron-Lavi) [inline webpack/assets/javascripts/react_app/redux/actions/toasts/index.js:6]: it was one annoying `uuid` ( especially in tests )

File: webpack/assets/javascripts/react_app/redux/actions/common/forms.js [hunk 3]
```diff
@@ -85,6 +79,7 @@ export const submitForm = ({
         actionTypes,
         handleError,
         handleSuccess,
+        successToast,
```
Comment (amirfefer) [inline webpack/assets/javascripts/react_app/redux/actions/common/forms.js:88]: does it make sense to add an error toast as well?

File: webpack/assets/javascripts/react_app/redux/API/APIActions.js [hunk 4]
```diff
@@ -10,8 +10,10 @@ const { GET, POST, PUT, DELETE, PATCH } = API_OPERATIONS;
  * @param { String } payload.url the url for the API request.
  * @param { String } payload.headers the API get request headers.
  * @param { Object } payload.params the API get request params.
- * @param { Object } payload.handleError an error handling callback.
- * @param { Object } payload.handleSuccess a success handling callback.
+ * @param { Function } payload.handleError an error handling callback.
+ * @param { Function } payload.handleSuccess a success handling callback.
+ * @param { String } payload.errorToast an error toast will be triggered with this message after API error.
+ * @param { String } payload.successToast a succes toast will be triggered with this message after API success.
```
Comment (amirfefer) [inline webpack/assets/javascripts/react_app/redux/API/APIActions.js]: can you add these new props to our documentation?

**Review by amirfefer — CHANGES_REQUESTED** [review]
Thanks @laviro ! a great addition to our API stack  :+1: 
can you please extend `APIRequest`  test for including this new feature?

**Comment by Ron-Lavi** [comment]
Thanks @amirfefer !
In the last commit:
- Added tests
- Added section to storybook
- Added also `errorToast` to `forms.js`
- Changed it into a function so the toast can contain info from the server, e.g:
```js
dispatch(
  get({
    key: MY_SPECIAL_KEY,
    url,
    toastError: error => `Oh no! Something went wrong while submiting the form, the server returned the following error: ${error}`,
    toastSuccess: response => `Yay! task: ${response.taskID} was successfully created.`
  })
);
```

**Review by amirfefer — APPROVED** [review]
Thanks @laviro ! 
works well

ready for merge after stabilization

**Comment by amirfefer** [comment]
Merging, Thanks @laviro !

**Pull request merged**